borg/platform_linux.pyx

```diff
@@ -141,3 +140,47 @@ def acl_set(path, item, numeric_owner=False):
                 acl_set_file(p, ACL_TYPE_DEFAULT, default_acl)
         finally:
             acl_free(default_acl)
+
+cdef _sync_file_range(fd, offset, length, flags):
+    if sync_file_range(fd.fileno(), offset, length, flags) != 0:
+        raise OSError(errno, os.strerror(errno))
+
+
+cdef request_sync(fd, offset, length):
+    _sync_file_range(fd, offset, length, SYNC_FILE_RANGE_WRITE)
+
+
+cdef force_sync(fd, offset, length):
+    _sync_file_range(fd, offset, length, SYNC_FILE_RANGE_WRITE | SYNC_FILE_RANGE_WAIT_BEFORE)
+    os.posix_fadvise(fd.fileno(), 0, 0, os.POSIX_FADV_DONTNEED)
+
+
+class SyncFile(BaseSyncFile):
+    """
+    Implemented using sync_file_range for asynchronous write-out and fdatasync for actual durability.
+
+    "write-out" means that dirty pages (= data that was written) are submitted to an I/O queue and will be send to
+    disk in the immediate feature.
+    """
+
+    def __init__(self, path):
+        self.fd = open(path, 'wb')
```

Comment:
Open/close is handled by LoggedIO, a context manager doesn't fit the code there. Note that the close() is inherited from platform_base.